src/halmos/mapper.py

```diff
@@ -0,0 +1,181 @@
+            raise ValueError(f"Contract {contract_name} not found")
+
+        contract_mapping_info.nodes.append(node)
+
+    def parse_ast(self, node: Dict, contract_name: str = ""):
+        node_type = node["nodeType"]
+
+        if node_type in self._PARSING_IGNORED_NODE_TYPES:
+            return
+
+        current_contract = self._get_current_contract(node, contract_name)
+
+        if node_type == "ContractDefinition":
+            if current_contract not in self._contracts:
+                self.add_contract_mapping_info(
+                    contract_name=current_contract, bytecode="", nodes=[]
+                )
+
+            if self.get_contract_mapping_info_by_name(current_contract).nodes:
+                return
+        elif node_type != "SourceUnit":
+            id, name, address, visibility = self._get_node_info(node, node_type)
+
+            self.append_node(
+                current_contract,
+                AstNode(node_type, id, name, address, visibility),
+            )
+
+        for child_node in node.get("nodes", []):
+            self.parse_ast(child_node, current_contract)
+
+        if "body" in node:
+            self.parse_ast(node["body"], current_contract)
+
+    def _get_node_info(self, node: Dict, node_type: str) -> Dict:
+        return (
+            node.get("id", ""),
+            node.get("name", ""),
+            "0x" + self._get_node_address(node, node_type),
+            node.get("visibility", ""),
+        )
+
+    def _get_node_address(self, node: Dict, node_type: str) -> str:
+        address_fields = {
+            "VariableDeclaration": "functionSelector",
+            "FunctionDefinition": "functionSelector",
+            "EventDefinition": "eventSelector",
+            "ErrorDefinition": "errorSelector",
+        }
+
+        return node.get(address_fields.get(node_type, ""), "")
+
+    def _get_current_contract(self, node: Dict, contract_name: str) -> str:
+        return (
+            node.get("name", "")
+            if node["nodeType"] == "ContractDefinition"
+            else contract_name
+        )
+
+    def find_nodes_by_address(self, address: str, contract_name: str = None):
+        if contract_name:
+            contract_mapping_info = self.get_contract_mapping_info_by_name(
+                contract_name
+            )
+
+            if contract_mapping_info:
+                for node in contract_mapping_info.nodes:
+                    if node.address == address:
+                        return node.name
+
+        result = ""
+        for key, contract_info in self._contracts.items():
+            matching_nodes = [
+                node for node in contract_info.nodes if node.address == address
+            ]
+
+            for node in matching_nodes:
+                result += f"{key}.{node.name} "
+
+        return result.strip() if result != "" and address != "0x" else address
+
+
+class DeployAddressMapper(metaclass=SingletonMeta):
```

Comment:
i'm concerned about DeployAddressMapper being a singleton object. unlike `Mapper` which is per project, `DeployAddressMapper` is per test. sharing a singleton across multiple tests don't seem like a good practice to me. also, i don't think this would would work with --test-parallel, due to potential race conditions. what do you think? cc: @karmacoma-eth

Reply:
good catch, I don't think DeployAddressMapper should be a singleton (or it should be scoped to a single test context). 

It seems to currently work with `--test-parallel` because it is not updated during the test run, it is only updated lazily in `render_trace` (so from the main thread). This likely means that values from a test can "leak" to another thread

Reply:
little repro:

```solidity
contract A {}

contract Test82 is Test, SymTest {
    function setUp() public {
        console2.log("Test82.setUp");
    }

    function test_DeployMapperLeak1_double_deploy() external {
        A a = new A();
        assert(address(a) != address(b));
    }

    function test_DeployMapperLeak2_deploys_nothing() external {
        vm.deal(address(this), 1 ether);
        payable(address(0xaaaa0002)).transfer(1 ether);
    }
}
```

if `deploys_nothing` runs first, it has the expected trace (`0xaaaa0002` is a naked address, not an A)

```
Trace:
    CALL Test82::test_DeployMapperLeak2_deploys_nothing
        CALL HEVM_ADDRESS::0xc88a5e6d00000000000000000000000000000000000000000000000000000000aaaa00010000000000000000000000000000000000000000000000000de0b6b3a7640000
        ↩ 0x
        CALL 0xaaaa0002::0x (value: 1000000000000000000)
        ↩ f_ret_256(call_exit_code_02())
    ↩ REVERT f_ret_256(call_exit_code_02()) (error: Revert())
```

If `double_deploy` runs first, the `DeployAddressMapper` is now polluted and the trace for `deploys_nothing` wrongly indicates that eth was transferred to an A instance:

```
Trace:
    CALL Test82::test_DeployMapperLeak2_deploys_nothing
        CALL HEVM_ADDRESS::0xc88a5e6d00000000000000000000000000000000000000000000000000000000aaaa00010000000000000000000000000000000000000000000000000de0b6b3a7640000
        ↩ 0x
        CALL A::0x (value: 1000000000000000000)    # wrong label here
        ↩ f_ret_256(call_exit_code_02())
    ↩ STOP 0x
```

Reply:
thanks for further clarification, karma!

if it is nontrivial to fix this, i'm ok with adding a todo comment, and opening a separate pr to fix it later.

Reply:
indeed, let's merge as-is and fix up later. I wanted to try a quick fix by introducing scopes for the singletons, but it's trickier than I thought